### PR TITLE
caml_sys_isatty: detect Cygwin/MSYS for better -color heuristic

### PR DESCRIPTION
**Pull request by nojb**
This PR improves the heuristic used to decide whether to use colors for warnings & errors by detecting the case when the native Windows port of OCaml is being used under Cygwin (which I would guess is fairly common).

@dra27 will know if we need to add some `#ifdef` to deal with the case of Windows XP and other strange creatures that may have trouble with `GetFileInformationByHandleEx`.

Of course, feel free to ignore this PR until after 4.06!

### CODE REVIEW

File: byterun/win32.c
Comment (gasche) [inline byterun/win32.c]: Just for my culture: the cygwin code you link to uses `L'\0'` instead of `0` to terminate widechar strings, are the two equivalent?

File: byterun/win32.c
Comment (gasche) [inline byterun/win32.c]: Personally I would find

```
  (wcsstr(.., L"msys-") || wcsstr(..., L"cygwin-"))
  && wcsstr(.., L"-pty")
```

more readable.

**Review by gasche — COMMENTED** [review]
Seems rather safe to me.

File: byterun/win32.c
Comment (nojb) [reply byterun/win32.c]: Yes.

File: byterun/win32.c
Comment (nojb) [reply byterun/win32.c]: Agree, will fix.

**Comment by dra27** [comment]
Not checked the detail, but I think you need to change the requested windows level (it's the VER define - I can't remember the precise name) to request Windows Vista. It would be advisable to load the symbol indirectly (as for CreateSymbolicLink, for example) - though at the moment I've dealing somewhat reactively rather than pre-emptively with Windows XP support!

**Comment by xavierleroy** [comment]
The intent of this PR sounds right, and I'm incompetent to comment on the implementation.  

A cosmetic suggestion: while you're at it, could you move the `_isatty` call to the `caml_win32_detect_msys_tty` function you're introducing, and rename that function to `caml_win32_isatty` or something clearer?  This way all the Windows-specific stuff (incl `_isatty`) is in file `win32.c`, as it should be.

**Comment by xavierleroy** [comment]
Also, I'm disturbed at the quantity of pull requests that suggest to do something for the Windows console and other terminals: the present PR, but also #1408 , and even #1399 .  Let's make sure we understand where we are going before merging all that stuff.

**Comment by dra27** [comment]
@xavierleroy - on "where we're going", #1399 I think was slightly misconceived (and has been closed as a result). The change here and in #1408 should be seen as trying to maintain parity between the OSes. This one moves towards consistent handling of coloured output (again, this trick is used in other projects) and there's another one in the pipeline - excuse pun - for the Windows Console itself. Amongst other things, #1408 brings feature parity for #1231, whose motivation was to allow UTF-8 to display characters instead of sequences.

**Comment by dra27** [comment]
I'd mentioned privately with @nojb, but I'll put it here for discussion. I'm wondering if it would be worth exposing this functionality in the Unix library. For Windows, there are two concepts for isatty:
 1. Some kind of interactive thing (so input coming from a user or output going to something a user can "see")
 2. Specifically a Windows Console

I'm wondering if we can add a function `Unix.isaconsole` which on Unix is just a (pointless) synonym for `Unix.isatty` but on Windows returns `true` strictly if the given descriptor is a Windows console. So we have `Unix.isaconsole` iff `Unix.isatty`.

**Comment by gasche** [comment]
Why would OCaml users care about the difference? Is the idea to move some of the console-specific logic currently in the runtime to userspace? (Or to make it easier for brave Windows-side maintainers to experiment outside the distribution?)

**Comment by nojb** [comment]
@xavierleroy Here is a less technical but maybe a little more conceptual explanation of what is going on and the reason for these Console-related PRs. Mainly directed at those who are not very familiar with how Windows works (like me; so apologies in advance for any inaccuracies!).

The first thing to keep in mind is that there is a fundamental difference between Unix and Windows in their handling of the "Console" (by which I mean what you get when you open up some kind of terminal and type things in it). For Unix systems there is basically no difference between writing & reading from the Console and writing & reading from a file: both are streams of bytes. However, in Windows the "Console" is not a stream of bytes but a **two dimensional array of (wide) characters**.

As a corollary, this means that anything that gets sent to the Console needs to be translated at some point into a sequence of wide characters, which means that 1) it *has* to have an encoding and 2) that encoding needs to be made known to the OS so that suitable re-encoding into wide characters (a.k.a. UTF-16) can take place. Regarding 1) #1200 has made it much more likely that whatever is being sent to the Console is UTF-8 encoded. Point 2) is typically done by calling the Windows API [`SetConsoleOutputCP`](https://docs.microsoft.com/en-us/windows/console/setconsoleoutputcp) or executing `chcp 65001` before starting your program. This **needs to be done** so that one can print UTF-8 strings using `write` or `printf`.

The problem is the **existence of several bugs in Windows** in the handling of the above situation, some of which have only been fixed rather recently.  This means that today, the native Windows port of `trunk` running on the Command Prompt (`cmd.exe`) has trouble writing and reading Unicode (and it can even segfault in one case).  Since it is rather sad to have modified the whole runtime system to support Unicode on Windows and then have it basically unusable through the Console, @dra27 put a lot of energy into working around these bugs. This is what #1408 is about.

One way to sidestep the need to deal with the encoding points above is to directly read and write wide characters.  Then no re-encoding is necessary and things will mostly just work. This is in a sense the most "native" way to talk to the Console on Windows and is done by calling [`ReadConsole`](https://docs.microsoft.com/en-us/windows/console/readconsole) and [`WriteConsole`](https://docs.microsoft.com/en-us/windows/console/writeconsole). It is by using these functions to implement the basic I/O primitives of the runtime system (for suitable versions of Windows) that #1408 works around the Windows bugs.

Lastly, it is also important to keep in mind that this discussion, and the modifications of #1408, are only relevant when writing & reading from the Console.  As soon as we redirect those handles into, say, files we are back into the land of streams of bytes and the situation is very much like in Unix. Note that this includes the case of the native Windows ports under Cygwin since Cygwin redirects those channels behind the scenes to implement their emulation layer. This means, in particular, that you can try out the native Windows port of `trunk` under Cygwin *today* and input/output of Unicode works perfectly (as perfectly as it works under Unix).

So, **TL; DR:**
- #1408 Works around Windows bugs in most (all?) versions of Windows so that I/O of Unicode works correctly on `cmd.exe`
- #1416 does "2)" (communicating the encoding to the OS) for those versions of Windows where this works correctly. 
- #1406 (this PR) improves the `-color` heuristic by detecting the situation of the native Windows port running under Cygwin, as described in the last paragraph. Not really related to any Unicode changes.

**Comment by xavierleroy** [comment]
Thanks, @nojb, for the high-level overview of the ongoing Win32 console work.  Now I can express my remaining concern more clearly.
* @dra27 recently implemented `Unix.isatty` for Win32 using `GetConsoleMode` and eschewing `_isatty`. 
* This PR is about `caml_sys_isatty`, a primitive that is not made available in the Caml libraries, not even as `Unix.isatty`.  It is using `_isatty` and you're adding other cases of console-like output.

So, we end up with two `isatty` functions whose behaviors and implementations have completely diverged. Isn't this a case of the right hand ignoring the left hand?  Couldn't we do better?

**Comment by dra27** [comment]
@nojb and my Slack history testifies that we definitely aren't ignoring each other :smile:

I haven't (yet) reviewed this PR in detail - the use of _isatty on Windows is a bug

File: byterun/sys.c
Comment (xavierleroy) [inline byterun/sys.c]: It looks like this URL is no longer relevant; why not delete the comment?

File: byterun/win32.c
Comment (xavierleroy) [inline byterun/win32.c]: Looks like this function could (should?) take a HANDLE as argument and not a file descriptor, saving a `_getosfhandle` call that is redundant with the one done in `caml_win32_isatty`.

File: byterun/win32.c
Comment (xavierleroy) [inline byterun/win32.c]: I had to read this twice to make sure the `&&` and `||` were in the right places.  What about:
```
if (hFile == INVALD_HANDLE_VALUE) return 0;
if (GetFileType(hFile) == FILE_TYPE_CHAR && GetConsoleMode(hFile, &lpMode)) return 1;
return caml_win32_ismsystty(h);
```

**Review by xavierleroy — APPROVED** [review]
Minor stylistic comments below.  Globally I'm in favor of merging.

File: byterun/sys.c
Comment (nojb) [reply byterun/sys.c]: Done.

File: byterun/win32.c
Comment (nojb) [reply byterun/win32.c]: Indeed, well spotted!

File: byterun/win32.c
Comment (nojb) [reply byterun/win32.c]: Yes, the suggested version is indeed clearer.

**Comment by nojb** [comment]
Amended as suggested by the reviewer.

Quick summary of recent changes:

I went ahead and merged the implementation of `Unix.isatty` under Windows and `caml_sys_isatty` (the primitive used to decide whether compiler colors should be used). Concretely, this means that `Unix.isatty` now returns `true` for the native Windows ports when passed a file descriptor connected to Cygwin's terminal.

Technically this is a breaking change (and is marked as such in `Changes`) but the doc string of `Unix.isatty` was always a bit ambiguous on this point (it reads: *Return true if the given file descriptor refers to a terminal or console window, false otherwise*).

**Comment by xavierleroy** [comment]
No need to worry about changing the semantics of `Unix.isatty` under Windows, because as you said the intended semantics has never been clear to begin with.

Under Unix, the traditional meaning of `isatty` is "supports the system calls for terminal I/O such as `tcsetattr` and the like". Historically it refers to serial lines; more recently, to pseudo-ttys.  The more modern meaning of `isatty` is "there is a terminal-style display and keyboard underneath, and a human user interacting with all that".  Under Windows, `tcsetattr` and the like are not implemented in the OCaml interface, and nobody cares about RS-232 serial lines any longer, so it's OK to stick to the modern meaning.

**Review by xavierleroy — APPROVED** [review]
Looks good to me!

**Comment by xavierleroy** [comment]
@dra27: would you like to review and give the final green light?

File: Changes
Comment (dra27) [inline Changes]: Possibly worth a tiny clarification to the text of this: it's a "file descriptor connected to *a Cygwin PTY*" and worth noting that it also means it works "under Cygwin */ MSYS2*"

File: byterun/win32.c
Comment (dra27) [inline byterun/win32.c]: Largely cosmetic, but MSYS is a fork of Cygwin, so I'd say "Detect if a named pipe corresponds to *a Cygwin/MSYS2 pty*"!

File: byterun/win32.c
Comment (dra27) [inline byterun/win32.c]: Either refer to the function name (`handle_to_fn`) or reference a specific commit - it's good to have a link to the source of this, but the file may change (in this case, s/master/00e9bf2 would solve it)

File: byterun/win32.c
Comment (dra27) [inline byterun/win32.c]: This needs a bounds check, however unlikely overflow may be.

File: byterun/win32.c
Comment (dra27) [inline byterun/win32.c]: Can this be `caml_win32_is_cygwin_pty` (both change to Cygwin, which I think is clearer as we don't "officially" support MSYS2, and also that we're really talking about PTYs, not TTYs here).

File: byterun/win32.c
Comment (dra27) [inline byterun/win32.c]: This could be explicitly "FILE_INFO_BY_HANDLE_CLASS and FILE_NAME_INFO are only available from Windows Vista onwards"

File: byterun/win32.c
Comment (dra27) [inline byterun/win32.c]: FWIW, the CreateSymbolicLink test updates a global (although for some reason I chose to use a second global there, rather than just initialising it to `INVALID_HANDLE_VALUE`) rather than attempting to locate the function each time https://github.com/ocaml/ocaml/blob/trunk/otherlibs/win32unix/symlink.c#L51-L55. But efficiency doesn't really matter for this test, I guess!

File: byterun/win32.c
Comment (dra27) [inline byterun/win32.c]: I don't particularly "like" the fact GetFileType is queried twice: why not do that once here:

```C
CAMLexport in caml_win32_isatty(int fd)
{
  DWORD lpMode;
  HANDLE hFile = (HANDLE)_get_osfhandle(fd);

  if (hFile == INVALID_HANDLE_VALUE)
    return 0;

  switch (GetFileType(hFile)) {
  case FILE_TYPE_CHAR:
    /* Both console handles and the NUL device are FILE_TYPE_CHAR,
       The NUL device returns FALSE for a GetConsoleMode call. _isatty
       incorrectly only uses GetFileType (see GPR#1321). */
    return GetConsoleMode(hFile, &lpMode);
  case FILE_TYPE_PIPE:
    /* Cygwin PTYs are implemented using named pipes */
    return caml_win32_is_cygwin_pty(hFile);
  default:
    return 0;
  }
}
```

File: byterun/win32.c
Comment (dra27) [inline byterun/win32.c]: Just for brevity, function pointer definitions don't need parameter names (cf. https://github.com/ocaml/ocaml/blob/trunk/otherlibs/win32unix/symlink.c#L31)

**Review by dra27 — CHANGES_REQUESTED** [review]
The only "important" change what I'm fairly sure is a missing bounds check.

#1321 fell foul of my usual trick of including information in commit messages which should be in code. If you choose not to refactor caml_win32_isatty as suggested, please do add comments explaining why we don't use `_isatty`

File: byterun/win32.c
Comment (nojb) [reply byterun/win32.c]: I am passing the maximum buffer size `(sizeof(buffer) - sizeof(WCHAR))` to the `GetFileInformationByHandleEx` call.  Do I understand correctly that you want the bounds check in case the system call returns something erroneous?

File: byterun/win32.c
Comment (nojb) [reply byterun/win32.c]: Good point; switched to a static variable.

File: byterun/win32.c
Comment (dra27) [reply byterun/win32.c]: Ah, I'm sorry - I'd missed the subtlety of the `- sizeof(WCHAR)` ... please could that be commented on the call (something like "GetFileInformationByHandleEx doesn't null-terminate the string, so reduce the declared buffer length to allow for adding one.")

File: Changes
Comment (dra27) [inline Changes]: Erroneous "the" at the end of the line!

File: byterun/win32.c
Comment (dra27) [inline byterun/win32.c]: Stray blank?

File: byterun/win32.c
Comment (dra27) [inline byterun/win32.c]: You could initialise this to `INVALID_HANDLE_VALUE`...

File: byterun/win32.c
Comment (dra27) [inline byterun/win32.c]: ... and compare that here instead of `NULL` (means that the lookup only happens once, even on Windows XP)... but it's up to you if you want to use keypresses making Windows XP execution "faster" :wink:

Comment (nojb) [reply byterun/win32.c]: Sure, done.

**Comment by nojb** [comment]
Thanks for the careful review @dra27! I think I addressed all the points raised.

**Review by dra27 — APPROVED** [review]

**Comment by dra27** [comment]
Lovely, thank you - let's wait for the CI to confirm everything's still in order and then merge (I presume you're happy for me to squash this one?)

**Comment by nojb** [comment]
Yes, sure. Thanks!

**Pull request merged**
